File: sgp.rb
#!/usr/bin/ruby -w
require 'digest'
require 'io/console'
=begin This is my Ruby version of SuperGenPass 
	 hash password generation. 2018 by mnl@vandal.nu
	 Written as an excercise. Outputs to stdout
=end
domain, len, $dig = ARGV

unless domain
	puts "Usage #$0 [domainname] [length (optional)]\n"; exit 1;
end
#Set password length to 15 is not user defined
len = (len.to_i.equal? 0) ? 15 : len.to_i

case $dig
  when "md5",nil then $dig = Digest::MD5 #md5 is default
  when "sha512" then $dig = Digest::SHA512
  else
	  puts "Unknown digest: #{$dig}"; exit 1
end

print "Password: "
master = STDIN.noecho(&:gets).chomp
hash = master+":"+domain

def valid(h)
	# Has to start with lowercase and contain uppercase and numbers
	if ( h =~ /^[a-z]/ and h =~ /[A-Z]/ and h =~ /[0-9]/ )
		return true
	end
end

10.times do |count|
	hash = $dig.base64digest(hash)
	hash.tr!('+/=','98A')
	next if count < 9 # Hash ten times and then until valid
	redo unless valid(hash)
end

#Done crop and print out
puts hash[0...len]

File: sgp.pl
#!/usr/bin/env perl
=begin comment
Perl SuperGenPass implementation. By mnl@vandal.nu 2018
Currently without url stripping. Prompt hiding only 
works on *nix with stty
=cut
use strict;
use warnings;
use Digest::SHA qw(sha512_base64);
use Digest::MD5 qw(md5_base64);
 
if ($#ARGV < 0) {
	print "Usage $0 [domainname] [length (optional)]\n"; exit 1;
}
system('/usr/bin/stty', '-echo');  # Disable echoing
print "Password: "; 
chomp(my $master = <STDIN>); 
system('/usr/bin/stty', 'echo');   # Turn it back on
my $len = $ARGV[1] ? $ARGV[1] : 15;
my $dig = $ARGV[2] ? $ARGV[2] : "md5";
my $hash ="$master:$ARGV[0]";
my $i = 1;

sub valid {
#SuperGenPass hashes should be hashed at least 10 times and until
# they start with [a-z], contain [A-Z] and [0-9]
	return (($i > 10 ) and $hash =~ m/^[a-z](?=.*\d)(?=.*[A-Z])/ ) 
}

until (valid) {
	if ($dig eq "sha512") {
		$hash = sha512_base64($hash); 	}
	elsif ($dig eq "md5" ) { 
		$hash = md5_base64($hash); }
	else { print "\nUnknown digest: $dig\n"; exit 1 }

	while (length($hash) % 4) {
		$hash .= 'A'; #perl's Digest doesn't add padding
	}
	$hash =~ tr!+/!98!; #SGP requires these chars translated
	$i++;
}

$hash =~ s/.{$len}\K.*//s; #Cut to max length
print "$hash\n";
